### PR TITLE
Add support for story mode and level photos

### PR DESCRIPTION
**Pull request by Slendy**
# Notable changes
* Story mode levels now support online functions like uploading and viewing scores, photos, comments, along with the ability to heart story levels.
* All levels will show photos taken from within the level as well.
* The website now includes info about the level a photo was taken in and level pages now show photos
* Hearted level lists and queue lists are now sorted by the time that you added them to the list
* The /users endpoint used to return an entire user's profile but packet captures show that it should only be returning the user's profile picture. The usage for the endpoint is when the game needs to find the profile pictures for lots of users like viewing levels and recent activity

# What needs more testing
* Do all games that we plan to support have unique story mode level ids?
* What happens if a story level from a previous game is hearted? What do people from a newer game see?
* If someone crafts their own authenticated post request to heart a story level with an invalid id, the hearted levels list will disappear from the user's profile because the invalid slot will get returned when pageSize=1. The current workaround is to only show hearted user levels in the preview box but this definitely needs a little bit more work.
* I double-checked my migration files this time but there's always a chance I messed those up

# Preview
In-game story level with scores, comments, and photos working
image

How levels look on the lighthouse website now
image

How photos look in the photos tab
image

Another example of a story mode level on the website
image

### CODE REVIEW

**Review by jvyden — CHANGES_REQUESTED** [review]
Good work!

It's a shame I didn't start tracking slot data in the initial photo implementation. Oh well. 

I'm also not sure about removing the foreign keys to slots from other models. Is there a problem I don't see with doing that? Because it seems like you could just get away with adding the slot type to the database and having the two types mix just fine. :thinking:

Also test failures seem to be actually relevant this time :upside_down_face:

**Comment by Slendy** [comment]
I tried to leave the foreign keys alone at first but whenever I would try and add a story level it would cause errors because it couldn't find a slot with that id

File: ProjectLighthouse/Helpers/SlotTypeHelper.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        return type switch
+        {
+            SlotType.Developer => "developer",
+            SlotType.User => "user",
+            SlotType.Pod => "pod",
+            _ => "unknown",
+        };
+    }
+
+    public static async Task<string> serializeDeveloperSlot(Database db, int id)
```
[the remark this answers is not in the document]
Comment (Slendy) [reply ProjectLighthouse/Helpers/SlotTypeHelper.cs]: The Java naming conventions are burned into my brain so I default to these lol.

**Comment by Slendy** [comment]
As for the unit tests, I think you need to look into a better way to generate random usernames because every time the registration test fails there is always one user that only has a '0' after their name.

**Comment by Slendy** [comment]
The website register tests were failing because the username field on the register page is limited to 16 characters. The test was generating usernames over the limit and the webdriver wasn't able to fully enter it. The password field was also being filled in with an already hashed value which would get hashed again by the website, but this didn't have any effect on the result.

**Review by jvyden — APPROVED** [review]
I'm going to take a look at the foreign keys issues after I wrap up implementing email verification, because that's something I do want.

For now though the code looks good, so I'll approve. I wont merge until I can resolve the issues with foreign keys, though.

**Comment by jvyden** [comment]
Alright, it's immediately clear to me that re-adding foreign keys is gonna be a task and a half, especially so since new things have been built upon the system that doesn't have them.

Here's what I'll do:

1. Close this pull request
2. Reimplement the changes (co-authoring you, of course), making sure to handle developer slots through the database therefore allowing foreign keys to be a thing.
3. Open a brand-new pull request with the changes
4. At that point, you can review the new changes just to double check
5. And then finally, we can merge.

In short, I'm nuking it and starting over :trollface:
